x/ecocredit/core/genesis.go

```diff
@@ -0,0 +1,267 @@
+package core
+
+import (
+	"context"
+	"encoding/json"
+
+	"github.com/cosmos/cosmos-sdk/codec"
+	"github.com/cosmos/cosmos-sdk/orm/model/ormdb"
+	"github.com/cosmos/cosmos-sdk/orm/model/ormtable"
+	"github.com/cosmos/cosmos-sdk/orm/types/ormjson"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	"github.com/gogo/protobuf/proto"
+	dbm "github.com/tendermint/tm-db"
+	"google.golang.org/protobuf/reflect/protoreflect"
+
+	api "github.com/regen-network/regen-ledger/api/regen/ecocredit/v1"
+	"github.com/regen-network/regen-ledger/types/math"
+	"github.com/regen-network/regen-ledger/x/ecocredit"
+)
+
+// Validate performs basic validation for each credit-batch,
+// it returns an error if credit-batch tradable or retired supply
+// does not match the sum of all tradable or retired balances
+func ValidateGenesis(data json.RawMessage, params Params) error {
+	if err := params.Validate(); err != nil {
+		return err
+	}
+
+	db := dbm.NewMemDB()
+	backend := ormtable.NewBackend(ormtable.BackendOptions{
+		CommitmentStore: db,
+		IndexStore:      db,
+	})
+
+	ormdb, err := ormdb.NewModuleDB(&ecocredit.ModuleSchema, ormdb.ModuleDBOptions{})
+	if err != nil {
+		return err
+	}
+
+	ormCtx := ormtable.WrapContextDefault(backend)
+	ss, err := api.NewStateStore(ormdb)
+	if err != nil {
+		return err
+	}
+
+	jsonSource, err := ormjson.NewRawMessageSource(data)
+	if err != nil {
+		return err
+	}
+
+	err = ormdb.ImportJSON(ormCtx, jsonSource)
+	if err != nil {
+		return err
+	}
+
+	abbrevToPrecision := make(map[string]uint32)
+	for _, ct := range params.CreditTypes {
+		abbrevToPrecision[ct.Abbreviation] = ct.Precision
+	}
+
+	cItr, err := ss.ClassInfoTable().List(ormCtx, api.ClassInfoPrimaryKey{})
+	if err != nil {
+		return err
+	}
+	defer cItr.Close()
+
+	// make sure credit type exist for class abbreviation in params
+	for cItr.Next() {
+		class, err := cItr.Value()
+		if err != nil {
+			return err
+		}
+
+		if _, ok := abbrevToPrecision[class.CreditType]; !ok {
+			return sdkerrors.ErrNotFound.Wrapf("credit type not exist for %s abbreviation", class.CreditType)
+		}
+	}
+
+	classIds := make(map[uint64]uint64) // map of projectID to classID
```

Comment:
i like the convention you used above (with abbrevToPrecision) of naming map type variables <key>To<Value>. could we also do that here?